File: leetcode/0001-1000/0001-0100/0011-0020/0016-three-sum-closest/ThreeSumClosest.ts
function threeSumClosest(numbers: number[], target: number): number {
  const n = numbers.length;

  let minDiff = Number.MAX_VALUE;
  let min = Number.MAX_VALUE;

  numbers.sort((a, b) => a - b);

  for (let i = 0; i < n - 2; i++) {
    const first = numbers[i];

    let j = i + 1;
    let k = n - 1;

    while (j < k) {
      const second = numbers[j];
      const third = numbers[k];

      const sum = first + second + third;

      const diff = Math.abs(sum - target);

      if (diff < minDiff) {
        minDiff = diff;
        min = sum;
      }

      if (sum < target) {
        j++;
      } else if (sum > target) {
        k--;
      } else {
        return target;
      }
    }
  }

  return min;
};

console.log(threeSumClosest([-1, 2, 1, -4], 1));

File: leetcode/0001-1000/0001-0100/0011-0020/0016-three-sum-closest/ThreeSumClosest.java
import java.util.Arrays;

public class ThreeSumClosest {

  public static void main(String[] args) {
    var solution = new Solution();

    System.out.println(
        solution.threeSumClosest(new int[] { -1, 2, 1, -4 }, 1));
  }
}

class Solution {
  public int threeSumClosest(int[] nums, int target) {
    Arrays.sort(nums);

    var minDiff = Integer.MAX_VALUE;
    var min = Integer.MAX_VALUE;
    var n = nums.length;

    for (var i = 0; i < n - 2; i++) {
      var first = nums[i];
      var j = i + 1;
      var k = n - 1;

      while (j < k) {
        var second = nums[j];
        var third = nums[k];
        var sum = first + second + third;

        if (minDiff > Math.abs(target - sum)) {
          minDiff = Math.abs(target - sum);
          min = sum;
        }

        if (sum == target) {
          return target;
        } else if (sum > target) {
          k--;
        } else {
          j++;
        }
      }
    }
    return min;
  }
}
